### PR TITLE
cmd-reference: document `import-db`

### PR DESCRIPTION
**Pull request by skshetry**

### CODE REVIEW

File: content/docs/command-reference/import-db.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+| **Oracle**          | `cx_Oracle`                       | `oracle://{username}:{password}@{hostname}/{database_name}`                                         |
+| **PostgreSQL**      | `psycopg2`                        | `postgresql://{username}:{password}@{hostname}/{database_name}`                                     |
+| **Snowflake**       | `snowflake-sqlalchemy`            | `snowflake://{user}:{password}@{account}.{region}/{database}?role={role}&amp;warehouse={warehouse}` |
+| **SQLite**          | -                                 | `sqlite://path/to/file.db?check_same_thread=false`                                                  |
+| **SQL Server**      | `pyodbc`                          | `mssql+pyodbc://{username}:{password}@{hostname}:{port}/{database_name}`                            |
+| **Trino**           | `trino`                           | `trino://{username}:{password}@{hostname}:{port}/{catalog}`                                         |
+
+DVC uses sqlalchemy internally. So DVC should support any SQL databases that
```
Comment (skshetry) [inline content/docs/command-reference/import-db.md:133]: I have only tested with `sqlite`, `postgres`, and `snowflake`. But it should probably work with all mentioned here (and more).

**Comment by dberenbaum** [comment]
Can we note that this is experimental and subject to breaking changes?

File: content/docs/command-reference/import-db.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+## Installing database drivers
+
+DVC does not come preinstalled with all the drivers for databases, you’ll need
+to install the required packages for the database you want to use.
+
+Some of the recommended packages are shown below, with their expected connection
+strings:
+
+| **Database**        | **PyPI package**                  | **Connection String**                                                                               |
```
Comment (dberenbaum) [inline content/docs/command-reference/import-db.md:72]: This is nice -- I didn't expect to document all the connection strings for different backends. Should we also include links to https://docs.sqlalchemy.org/en/20/core/engines.html#backend-specific-urls and/or https://docs.sqlalchemy.org/en/20/dialects/index.html?

File: content/docs/command-reference/import-db.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+| **Oracle**          | `cx_Oracle`                       | `oracle://{username}:{password}@{hostname}/{database_name}`                                         |
+| **PostgreSQL**      | `psycopg2`                        | `postgresql://{username}:{password}@{hostname}/{database_name}`                                     |
+| **Snowflake**       | `snowflake-sqlalchemy`            | `snowflake://{user}:{password}@{account}.{region}/{database}?role={role}&amp;warehouse={warehouse}` |
+| **SQLite**          | -                                 | `sqlite://path/to/file.db`                                                                          |
+| **SQL Server**      | `pyodbc`                          | `mssql+pyodbc://{username}:{password}@{hostname}:{port}/{database_name}`                            |
+| **Trino**           | `trino`                           | `trino://{username}:{password}@{hostname}:{port}/{catalog}`                                         |
+
+DVC uses `sqlalchemy` internally. So DVC should support any SQL databases that
+provide dialects for SQLAlchemy. Refer to their documentation for more details.
```
Comment (dberenbaum) [inline content/docs/command-reference/import-db.md:86]:
```suggestion
provide dialects for SQLAlchemy. Refer to their [documentation](https://docs.sqlalchemy.org) for more details.
```

File: content/docs/command-reference/import-db.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+## Database Permissions
+
+It is recommended to use an user account with limited access to databases and
+tables and one that only has minimal read-only privileges like running `SELECT`
+queries. Different databases have different approaches to this. Refer to their
+documentation for more details.
```
Comment (dberenbaum) [inline content/docs/command-reference/import-db.md:93]: Should we say briefly why it's important? Also, minor, but WDYT of putting this into an admon/warning to call it out?

File: content/docs/command-reference/import-db.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+- `-v`, `--verbose` - displays detailed tracing information.
+
+## Examples
+
+### Downloading sql query result
+
+To import from a database using a `db` config set:
+
+```dvc
+$ dvc import-db --sql "select * from customers" --conn pgsql
```
Comment (dberenbaum) [inline content/docs/command-reference/import-db.md:130]: Do we need to do something to the syntax highlighter to support new commands @yathomasi?

File: content/docs/command-reference/import-db.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+documentation for more details.
+
+## Options
+
+- `-o <path>`, `--out <path>` - specify a `path` to the desired location in the
+  workspace to place the file. The default filename will be `results`, with an
+  extension name added based on `--output-format`.
+
+- `--sql <query>` - SQL query to run. You can also use a connection string set
+  in config, with `--conn`.
```
Comment (dberenbaum) [inline content/docs/command-reference/import-db.md:102]: > You can also use a connection string set
  in config, with `--conn`.

I don't follow what this means or why it's mentioned in `--sql`.

File: content/docs/user-guide/project-structure/configuration.md [hunk 7]
```diff
@@ -215,6 +217,33 @@ section):
...
+## db
+
+Similar to `remote`, configuration files may have more than one `'db'`. All of
+them require a unique `"name"` and a `url` value which is a connection string to
+connect to the database. They can also specify `username` and `password`
+options, which is used to combine with provided `url`, which is what is passed
+to the appropriate database drivers to connect to the database.
+
+For security reasons, it is recommended not to add password in the url, but
+instead set it separately in a local file.
```
Comment (dberenbaum) [inline content/docs/user-guide/project-structure/configuration.md:229]: Do you mean using `dvc config --local` or somewhere else not managed by DVC?

**Review by dberenbaum — CHANGES_REQUESTED** [review]
Left some comments on the doc, but it raised a bigger product question for me that I missed in https://github.com/iterative/dvc/pull/10125. Unlike https://github.com/iterative/dvc/pull/10040, `--sql` is required here, right? It feels artificial to be limited to a sql query written in the command line. Users likely either want to pull a table, which makes `select *` seem redundant, or the results of a more complex query.

**Comment by skshetry** [comment]
> Left some comments on the doc, but it raised a bigger product question for me that I missed in [iterative/dvc#10125](https://github.com/iterative/dvc/pull/10125). Unlike [iterative/dvc#10040](https://github.com/iterative/dvc/pull/10040), `--sql` is required here, right? 

We discussed this privately around the time of iterative/dvc#10040 about snapshotting a table or a database, but at that point `--sql` was not the main focus. It should be easy to add a `--table` here.

File: content/docs/command-reference/import-db.md (hunk 5 above)
Comment (yathomasi) [reply content/docs/command-reference/import-db.md:130]: Yes, we need to add it to https://github.com/iterative/gatsby-theme-iterative/blob/main/packages/gatsby-theme-iterative/config/prismjs/dvc-commands.js

File: content/docs/user-guide/project-structure/configuration.md (hunk 7 above)
Comment (skshetry) [reply content/docs/user-guide/project-structure/configuration.md:229]: Yup, I mean `--local`. This negative is probably confusing, so I'll reword this.

File: content/docs/command-reference/import-db.md (hunk 6 above)
Comment (skshetry) [reply content/docs/command-reference/import-db.md:102]: Leftover from dbt.

File: content/docs/command-reference/import-db.md (hunk 3 above)
Comment (skshetry) [reply content/docs/command-reference/import-db.md:86]: With "their", I mean the particular plugins for the database.

File: content/docs/command-reference/import-db.md (hunk 2 above)
Comment (skshetry) [reply content/docs/command-reference/import-db.md:81]: I have added a link to the former one.

File: content/docs/command-reference/import-db.md (hunk 4 above)
Comment (skshetry) [reply content/docs/command-reference/import-db.md]: I have moved this to the top and added an admon (but there are 2 back to back admon).

**Comment by dberenbaum** [comment]
Depends on https://github.com/iterative/dvc/pull/10219

**Comment by dberenbaum** [comment]
> Can we note that this is experimental and subject to breaking changes?

@skshetry Can we add something like this? Otherwise it LGTM.

**Comment by skshetry** [comment]
> @skshetry Can we add something like this? Otherwise it LGTM.

Should we consider it stable now?

**Comment by dberenbaum** [comment]
I would rather play it safe here, especially considering the discussions in https://github.com/iterative/dvc/pull/10164 and related issues that could impact this. We don't do this so often that it's hard to maintain, so I don't see much downside.

File: content/docs/command-reference/import-db.md [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+```usage
+usage: dvc import-db [-h] [-q | -v]
+            [--table TABLE] [--sql SQL] [--conn [CONNECTION]]
+            [--output-format [{csv,json}]] [-o [<path>]] [-f]
+```
+
+## Description
+
+With `import-db`, you can snapshot your ETL/database to a file to use in your
+data pipelines. This commands supports importing your table or a sql query
```
Comment (shcheklein) [inline content/docs/command-reference/import-db.md:14]: sql -> SQL (consistency)

File: content/docs/command-reference/import-db.md [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+results into different file formats. To do so, you have to set connection
+strings to connect to a database, which can be setup in config as `db.<name>`.
+Check [Database Connections](#database-connections) for more information.
+
+At the moment, `import-db` supports two different output format:
+
+- json records
```
Comment (shcheklein) [inline content/docs/command-reference/import-db.md:21]: JSON
CSV

File: content/docs/command-reference/import-db.md [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+<admon type="info">
+
+You can `dvc push` and `dvc pull` data imported from the databases to/from
+remote storage normally.
+
+</admon>
+
+## Database Connections
+
+To connect to a database, dvc needs a database connection string URI. This has
```
Comment (shcheklein) [inline content/docs/command-reference/import-db.md:37]: DVC

File: content/docs/command-reference/import-db.md [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+to be configured in the [`db`] section.
+
+```dvc
+dvc config db.pgsql.url postgresql://user@hostname:port/database
```
Comment (shcheklein) [inline content/docs/command-reference/import-db.md:41]: start with $ - both commands

File: content/docs/command-reference/import-db.md [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+as `--sql` can run arbitrary queries. Different databases have different
+approaches to this. Refer to their documentation for more details.
+
+</admon>
+
+You need to specify the name of database connection to use, when using
+`import-db`.
+
+```dvc
+dvc import-db --table customers_table --conn pgsql
```
Comment (shcheklein) [inline content/docs/command-reference/import-db.md:64]: start with $

File: content/docs/command-reference/import-db.md [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+## Options
+
+- `-o <path>`, `--out <path>` - specify a `path` to the desired location in the
+  workspace to place the file. If not specified, the filename will be generated
+  using the arguments from `--output-format` and `--table`, or for `--sql`, it
+  starts with "results" by default.
+
+- `--table <table>` - Table to snapshot.
+
+- `--sql <query>` - Execute SQL query and snapshot its result.
```
Comment (shcheklein) [inline content/docs/command-reference/import-db.md:108]: execute (starts with lower case)

File: content/docs/command-reference/import-db.md [hunk 14]
```diff
@@ -0,0 +1,172 @@
+# import-db
+
+Snapshot a table or a SQL query results from a database into CSV/JSON format.
+
+```usage
+usage: dvc import-db [-h] [-q | -v]
+            [--table TABLE] [--sql SQL] [--conn [CONNECTION]]
```
Comment (shcheklein) [inline content/docs/command-reference/import-db.md:7]: CONNECTION -> <connection>? (and below as well)

**Review by shcheklein — COMMENTED** [review]
https://dvc.org/doc/user-guide/project-structure/dvc-files - do we need to update some of these files?

also this https://dvc.org/doc/command-reference/update

**Comment by dberenbaum** [comment]
> https://dvc.org/doc/user-guide/project-structure/dvc-files - do we need to update some of these files?
> 
> also this https://dvc.org/doc/command-reference/update

Yes, good catch

**Comment by skshetry** [comment]
> [dvc.org/doc/user-guide/project-structure/dvc-files](https://dvc.org/doc/user-guide/project-structure/dvc-files) - do we need to update some of these files?
> 
> also this [dvc.org/doc/command-reference/update](https://dvc.org/doc/command-reference/update)

Since we consider this as experimental, I'd avoid updating other files for now.

**Comment by dberenbaum** [comment]
@skshetry Could you resolve the rest of the comments so we can merge?

**Comment by shcheklein** [comment]
> Since we consider this as experimental, I'd avoid updating other files for now.

could you folks share some light on this idea? what is the expectation to make a proper feature?

also, can we just put a note / admon on the top of the cmd ref pages? it would be more explicit I guess 

(not a blocker - just the first time I see that we are being conservative with updates to docs and just curios about the procedure you have in mind)

**Comment by skshetry** [comment]
> could you folks share some light on this idea? what is the expectation to make a proper feature?

See https://github.com/iterative/dvc.org/pull/5033#issuecomment-1877490708.

We want to see if we can fit this into `datasets` or `datasets`-like interface, since users are likely to read from their database in their scripts rather than snapshotting into a csv file.
But I don't think it is either-or here. (But snapshotting users may be smaller than streaming).

File: content/docs/command-reference/import-db.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+This is an experimental command, and is subject to breaking changes.
+
+</admon>
+
+Snapshot a table or a SQL query results from a database into CSV/JSON format.
+
+```usage
+usage: dvc import-db [-h] [-q | -v]
+            [--sql SQL | --table TABLE] [--conn CONN]
```
Comment (skshetry) [inline content/docs/command-reference/import-db.md:13]: Ideally, this should be:
```suggestion
            (--sql SQL | --table TABLE) --conn CONN
```

But auto-linking seems to be broken with this.

**Comment by skshetry** [comment]
> That's fine with me as long as we have an issue to track it so we don't forget.

Created https://github.com/iterative/dvc.org/issues/5087.

File: content/docs/command-reference/import-db.md (hunk 15 above)
Comment (shcheklein) [reply content/docs/command-reference/import-db.md:13]: I think we should change it in the DVC itself also. We don't use this style. I mean all uppercase CONN or TABLE. Pleas check other commands.

If this `(--sql SQL | --table TABLE)` is correct (why?) - we should also fix first on the DVC side - and adjust auto-linker here, I hope it's not that complicated

but let's check if we do this in some other DVC commands

Comment (skshetry) [reply content/docs/command-reference/import-db.md:13]: They are a mutually-exclusive argument, and one of them is required. I can change the casings of `TABLE`/`SQL` with metavar (uppercased name is just default, and we do those in many places when we are lazy).

Comment (skshetry) [reply content/docs/command-reference/import-db.md]: I have changed this in the docs. Will handle this in DVC separately.

Comment (shcheklein) [reply content/docs/command-reference/import-db.md]: >  and we do those in many places when we are lazy

are there examples? I checked a few commands and did't find it.

anyways, I think we need to update it on the DVC side for all the commands

> They are a mutually-exclusive argument, and one of them is required.

good point, then let's change on the DVC side first and we can fix the website later ... DVC is more important I think

Comment (skshetry) [reply content/docs/command-reference/import-db.md]:
```
usage: dvc experiments run [-h] [-q | -v] [-f] [-i] [-s] [-p] [-P] [-R] [--downstream] [--force-downstream] [--pull]
                           [--allow-missing] [--dry] [-k] [--ignore-errors] [-n <name>] [-S [<filename>:]<param_name>=<param_value>]
                           [--queue] [--run-all] [-j <number>] [--temp] [-C COPY_PATHS] [-m MESSAGE]
                           [targets ...]
```

See [-m MESSAGE] and [-c COPY_PATHS]. It's just the default from argparse, so it's easy to miss here. There are other commands like `du`/`artifacts get`/`exp save`/`fetch`/`ls-url`/`ls`/`get`/`import`, etc that does this. :)

Comment (skshetry) [reply content/docs/command-reference/import-db.md]: Fixed this in dvc by https://github.com/iterative/dvc/pull/10226.

**Review by dberenbaum — APPROVED** [review]
Thanks @skshetry! LGTM.

**Pull request merged**